### PR TITLE
use lineax to solve linear system in implicit diff

### PR DESCRIPTION
**Pull request by marcocuturi**
- use lineax `CG` (symmetric case) or `NormalCG` (not symmetric case) to handle linear solves, when possible (at this moment this hinges mostly on whether is running python >=3.9). Default back to JAX's native `scipy.sparse.linalg.cg` (or normal CG in unsymmetric case) when lineax cannot be imported.
- ridge parameters in linear solve can still be used for JAX solvers (`ridge_kernel` and `ridge_identity`) but are not handled for lineax solvers.
- fix hessians notebook: 
  - remove reference to identity preconditioning (everything uses default parameters now)
  - add check of eigenvalues of hessian and numerical accuracy of hessian approx.
- some tests are now only tested with `lineax`.

### CODE REVIEW

**Comment by michalk8** [comment]
@marcocuturi the failing tests can be remedied by adding an [importorskip](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-importorskip), e.g., inside the test functions that use `lineax`, I'd do `_ = pytest.importorskip("lineax")  # some comment explaining while this is needed`.

**Comment by marcocuturi** [comment]
> @marcocuturi the failing tests can be remedied by adding an [importorskip](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-importorskip), e.g., inside the test functions that use `lineax`, I'd do `_ = pytest.importorskip("lineax") # some comment explaining while this is needed`.

I went for something in the middle, sometimes I do use the ridge parameters so that JAX solvers run.

File: src/ott/solvers/linear/lineax_implicit.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    symmetric: whether `lin` is symmetric.
+    nonsym_solver: `lineax` solver used when handling non symmetric cases. Note
+      that `CG` is used by default in the symmetric case.
+    kwargs: arguments passed to `lineax`'s linear solver.
+  """
+  input_structure = jax.eval_shape(lambda: b)
+  kwargs.setdefault("rtol", 1e-6)
+  kwargs.setdefault("atol", 1e-6)
+  # Ridge parameters passed to JAX solvers are ignored in lineax.
+  _ = kwargs.pop("ridge_identity", None)
```
Comment (michalk8) [inline src/ott/solvers/linear/lineax_implicit.py:70]: Why do these need to be popped? They are no default arguments in `kwargs`; if it's just because of the tests, would remove the `kwargs.pop()` here and adjust the tests instead.

File: src/ott/solvers/linear/implicit_differentiation.py [hunk 2]
```diff
@@ -295,3 +283,42 @@ def gradient(
...
+    lin_t: Optional[LinOp_t] = None,
+    symmetric: Optional[bool] = False,
+    ridge_identity: float = 0.0,
+    ridge_kernel: float = 0.0,
+    **kwargs: Any
+) -> jnp.ndarray:
+  """Wrapper around JAX native linear solvers.
+
+  Args:
+    lin: Linear operator
```
Comment (michalk8) [inline src/ott/solvers/linear/implicit_differentiation.py:300]: Would do: Linear operator :math:`A` (also in the other function).

File: src/ott/solvers/linear/implicit_differentiation.py [hunk 3]
```diff
@@ -295,3 +283,42 @@ def gradient(
...
+    b: vector such that sought `x` is such that `lin(x)=b`
```
Comment (michalk8) [inline src/ott/solvers/linear/implicit_differentiation.py:301]: Too much "such that", would rephrase as:
Vector :math:`b` such that :math:`A x = b`.

And update the docs of the other function as well.

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: src/ott/solvers/linear/implicit_differentiation.py [hunk 4]
```diff
@@ -51,17 +67,18 @@ class ImplicitDiff:
       theorem differentiation.
   """
 
-  solver_fun: Callable[[jnp.ndarray, jnp.ndarray],
-                       Tuple[jnp.ndarray, ...]] = jax.scipy.sparse.linalg.cg
-  ridge_kernel: float = 0.0
-  ridge_identity: float = 0.0
+  solver: Optional[Solver_t] = None
+  solver_kwargs: Optional[Dict[str, Any]] = None
```
[the remark this answers is not in the document]
Comment (marcocuturi) [reply src/ott/solvers/linear/implicit_differentiation.py:71]: I think avoiding closure/partial is a bit preferable here. 

But maybe not clean because IIUC there's no way to mark a Callable that takes optional arguments (...). Another option would be to pass a dictionary (last input = Any) and "fish" variables in there?

File: src/ott/solvers/linear/implicit_differentiation.py [hunk 5]
```diff
@@ -295,3 +283,42 @@ def gradient(
 
   def replace(self, **kwargs: Any) -> "ImplicitDiff":  # noqa: D102
     return dataclasses.replace(self, **kwargs)
+
+
+def solve_jax_cg(
+    lin: LinOp_t,
+    b: jnp.ndarray,
+    lin_t: Optional[LinOp_t] = None,
+    symmetric: Optional[bool] = False,
```
[the remark this answers is not in the document]
Comment (marcocuturi) [reply src/ott/solvers/linear/implicit_differentiation.py:292]: Yes, thanks!

File: src/ott/solvers/linear/implicit_differentiation.py (hunk 3 above)
Comment (marcocuturi) [reply src/ott/solvers/linear/implicit_differentiation.py:301]: added something a bit less mathematical! (if we want to be math rigorous, it should be `A(x)`)

File: src/ott/solvers/linear/implicit_differentiation.py [hunk 6]
```diff
@@ -295,3 +283,42 @@ def gradient(
...
+    b: vector such that sought `x` is such that `lin(x)=b`
+    lin_t: Linear operator, corresponding to transpose of `lin`.
+    symmetric: whether `lin` is symmetric.
+    ridge_kernel: promotes zero-sum solutions. Only use if `tau_a = tau_b = 1.0`
+    ridge_identity: handles rank deficient transport matrices (this happens
+      typically when rows/cols in cost/kernel matrices are collinear, or,
+      equivalently when two points from either measure are close).
+    kwargs: arguments passed to jax.scipy.sparse.linalg.cg
+  """
+  l = lin if symmetric else lambda x: lin_t(lin(x))
```
[the remark this answers is not in the document]
Comment (marcocuturi) [reply src/ott/solvers/linear/implicit_differentiation.py:310]: thanks!

File: src/ott/solvers/linear/lineax_implicit.py (hunk 1 above)
Comment (marcocuturi) [reply src/ott/solvers/linear/lineax_implicit.py:70]: all right!
